GliaWidgets/SecureConversations/SecureConversations.UnreadMessageLoader.swift

```diff
@@ -0,0 +1,96 @@
+            let unreadCountRequestSource = ReactiveSwift.Signal<Result<Int, Error>, Never>.pipe()
+            let unreadCountTimeoutSource = ReactiveSwift.Signal<Result<Int, Error>, Never>.pipe()
+            // In case if `unreadCount` will never be delivered,
+            // (for example if sockets are disconnected, because
+            // `unreadCount` is based on sockets)
+            // we will still receive callback by timeout.
+            let unreadCountRequestWithTimeoutSource = ReactiveSwift.Signal.merge(
+                unreadCountRequestSource.output,
+                unreadCountTimeoutSource.output
+            ).take(first: 1)
+
+            let messagesSource = ReactiveSwift.Signal<Result<[ChatMessage], CoreSdkClient.SalemoveError>, Never>.pipe()
+
+            let combined = ReactiveSwift.Signal.zip(
+                messagesSource.output,
+                unreadCountRequestWithTimeoutSource
+            )
+
+            unreadCountTimeoutProducer
+                .delay(Self.unreadCountFallbackTimeoutSeconds, on: environment.scheduler)
+                .startWithCompleted {
+                    unreadCountTimeoutSource.input.send(value: .failure(TimeoutError()))
+                }
+
+            combined
+                .observe(on: environment.scheduler)
+                .observeValues { messageResult, unreadCountResult in
+                    callback(
+                        Self.messageWithUnreadCountResult(
+                            messageResult: messageResult.mapError { $0 as Error },
+                            unreadCountResult: unreadCountResult
+                        )
+                    )
+                }
+
+            environment.getSecureUnreadMessageCount(unreadCountRequestSource.input.send(value:))
+            environment.fetchChatHistory(messagesSource.input.send(value:))
+        }
+
+        static func messageWithUnreadCountResult(
+            messageResult: Result<[ChatMessage], Error>,
+            unreadCountResult: Result<Int, Error>) -> Result<MessagesWithUnreadCount, Error> {
+                // Without messages unread count does not make much sense,
+                // that is why we prefer to report error for messages in case
+                // of failure for both requests. Same for success - ignore
+                // unreadCount failure in case of successful loading of messages.
+                switch (messageResult, unreadCountResult) {
+                case let (.success(messages), .success(unreadCount)):
+                    return .success(MessagesWithUnreadCount(messages: messages, unreadCount: unreadCount))
+                case let (.success(messages), .failure):
+                    return .success(MessagesWithUnreadCount(messages: messages, unreadCount: .zero))
+                case let (.failure(error), .failure):
+                    return .failure(error)
+                case let (.failure(error), .success):
```

Comment:
Are these cases split for better readability? Because they both return the same regardless of what the second value is.

Reply:
Seems like indeed those can be combined for brevity. 935ba0b72b25c657b3d8cce6f03fd05ed8cc5414